keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,659 @@
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The `v1.Endpoints` API has been essentially (though not _actually_)
+deprecated since EndpointSlices became GA in 1.21. Several new Service
+features (such as dual-stack and topology, not to mention "services
+with more than 1000 endpoints") are implemented only for
+EndpointSlice, not for Endpoints. Kube-proxy no longer uses Endpoints
+ever, for anything, and the Gateway API conformance tests also require
+implementations to use EndpointSlices, so Gateway implementations
+don't use Endpoints either.
+
+Despite this, kube-controller-manager still does all of the work of
+managing Endpoints objects for all Services, and a cluster cannot pass
+the conformance test suite unless the Endpoints and EndpointSlice
+Mirroring controllers are running, even though in many cases nothing
+will ever look at the output of the Endpoints controller (and the
+EndpointSlice Mirroring controller will never output anything).
+
+While Kubernetes's API guarantees make it essentially impossible to
+ever actually fully remove Endpoints, we should at least move toward a
+world where most users run Kubernetes with the Endpoints and
+EndpointSlice Mirroring controllers disabled.
+
+## Motivation
+
+### Goals
+
+- Officially declare v1.Endpoints to be deprecated. Update
+  documentation and put out appropriate communications (blog posts,
+  etc) to ensure that end users are aware of this deprecation.
+
+- Add warnings (e.g., `Warning:` headers on Endpoints create/update)
+  to alert users of the fact that Endpoints is deprecated.
+
+- Update the e2e test suite to make it possible to run it in a "no
+  Endpoints controller" configuration, by rewriting some tests and
+  adding feature tags to others.
+
+- Update the conformance test suite to not require the Endpoints and
+  EndpointSlice Mirroring controllers to be running, by rewriting some
+  tests and demoting others from conformance.
```

Comment:
Conformance and e2e should be behavior driven so this is ok if the behavior is covered ... IIRC most of the tests use the endpoints/endpoint slices objects to synchronize the test and avoid races and flakes because of that, so a conformance test that uses these helpers should be ok to use just EndpointSlices